working/0723-static-extensions/feature-specification.md

```diff
@@ -315,165 +453,482 @@ extension.
 
-If an extension declaration _D_ named `E` declares a non-generic
-constructor _D1_ and the on-declaration of _D_ is generic then _D1_ is
-treated as a generic constructor that declares exactly the same type
-parameters as _D_, and it passes exactly the same actual type arguments to
-the constructor return type as the ones that are passed to the
-on-declaration in the on-type.
-
-*For example:*
-
-```dart
-extension E1<X, Y> on C<X, List<Y>, int> {
-  C.name(X x, Iterable<Y> ys): this(x, ys, 14);
-  // The previous line has the same meaning as the next line:
-  C<X, List<Y>, int>.name<X, Y>(X x, Iterable<Y> ys): this(x, ys, 14);
-}
-
-extension E2<X, Y> on D { // D is non-generic.
-  D.new(X x, Y y);
-  // Same as:
-  D.new<X, Y>(X x, Y y);
-}
-```
-
-#### Resolution of a constructor in an extension
+If an extension declaration declares a factory constructor, then the
+downwards context for the purposes of inference of the body of the
+constructor (or for inferring the arguments to the redirectee in the
+case of a redirecting factory constructor) is the on-type of the
+extension.
 
-Assume that `E` denotes an extension declaration _D_ with on-declaration
-_D1_ named `C`, and assume that _D_ declares a constructor whose name is
-`C`.
+*The on-type is well-formed within the scope of the body, since the
+generic type parameters of the extension are in scope in the
+constructor.*
 
-In that case an invocation of the form `E.new<TypeArguments>(arguments)` or
-the form `E.new(arguments)` is a fully resolved invocation of said
-constructor declaration.
+If an extension declaration declares a redirecting generative
+constructor, then the type arguments which are passed to the target of
+the redirection are the type arguments of the on-type of the
+extension.
 
-Similarly, if _D_ declares a constructor whose name is `C.name` then an
-invocation of the form `E.name<TypeArguments>(arguments)` or
-`E.name(arguments)` is a fully resolved invocation of said constructor
-declaration.
+*Unlike a redirecting generative constructor declared in a class, such
+a constructor declared in an extension may impose additional
+requirements on the actual type arguments provided to the target of
+the redirection.*
+
+It is a compile-time error if a redirecting factory constructor
+declared in an extension has a redirection target which is not
+assignable to the on-type of the extension (after inference has been
+performed).
+
+In addition, all of the usual compile-time errors associated with
+declarations of redirecting factory constructors continue to apply.
+
+*Inference is performed to reconstruct the type arguments, if any, for
+the target of the redirecting factory constructor, and the target of
+the redirection is checked for errors as usual.  In addition, the
+post-inference return type of the target of the redirection must be a
+type which is compatible with the return type of the constructor being
```

Comment:
compatible with -> a subtype of

Still no need to suggest anything less that this requirement.

We do need to be a little careful now, because we used to be able to say "must be a constructor on a subtype", but now you can forward directly to a constructor of an extension, so looking at the return type may be better. Or we should introduce a notion of "the constructed type" for a constructor, which is the declaring class/etc. for a non-extension constructor, and the `on` type for an extension constructor (instantiated or uninstantiated depending on whether it's from outside or inside the extension).